src/bottom_up_iterators/bottom_up_iterator.jl

```diff
@@ -0,0 +1,151 @@
+"""
+	abstract type BottomUpIterator <: ProgramIterator
+
+Enumerates programs in a bottom-up fashion. This means that complete programs are generated based on a user-provided order (e.g. depth, size, other cost function).
+
+Concrete implementations of this iterator should design the following custom data structures:
+- `bank::BottomUpBank`: Meant to store the list of already-generated programs.
+- `data::BottomUpData`: Any additional information the iterator might require.
+
+Additionally, the following functions should be implemented:
+- `init_bank(iter::BottomUpIterator)::BottomUpBank`: Returns an initialized `BottomUpBank`.
+- `init_data(iter::BottomUpIterator)::BottomUpData`: Returns an initialized `BottomUpData`.
+- `create_program!(iter::BottomUpIterator, bank::BottomUpBank, data::BottomUpData)::Union{Nothing, RuleNode}`: Generates the next program. Returns `nothing` if no other programs can be generated.
+- `update_state!(iter::BottomUpIterator, bank::BottomUpBank, data::BottomUpData, program::RuleNode)::Nothing`: Called after a program is validated.
+
+Observational equivalence (i.e. removing programs which yield the same outputs on all examples) is an optional optimization which could be used by `BottomUpIterator`s.
+To enable it, add the following fields to the `BottomUpIterator` implementation:
+- `problem::Union{Nothing, Problem{Vector{IOExample}}}.
+- `obs_equivalence::Bool` (and set this to `true` when creating the `BottomUpIterator`).
```

Comment:
I think this would ideally be a `HasObsEquivalence` trait instead, but if this approach works, I'm fine leaving it as is. It just makes the code for `_has_obs_equivalence` not-so-pretty.